File: comfyui/ui/src/schemas/physics/physics-schema.ts
/**
 * Physics Schemas
 *
 * Zod schemas for physics simulation types.
 * All numeric values use .finite() to reject NaN/Infinity.
 * Includes comprehensive validation constraints for security and data integrity.
 */

import { z } from "zod";
import { finiteNumber, normalized01, positiveFinite, nonNegativeFinite } from "../layers/primitives-schema";
import {
  entityId,
  boundedArray,
  MAX_ARRAY_LENGTH,
} from "../shared-validation";

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// Physics Enums
// ═══════════════════════════════════════════════════════════════════════════

export const BodyTypeSchema = z.enum([
  "static",
  "dynamic",
  "kinematic",
  "dormant",
  "AEmatic",
  "dead",
]);

export type BodyType = z.infer<typeof BodyTypeSchema>;

export const JointTypeSchema = z.enum([
  "pivot",
  "spring",
  "distance",
  "piston",
  "wheel",
  "weld",
  "blob",
  "rope",
]);

export type JointType = z.infer<typeof JointTypeSchema>;

export const ForceTypeSchema = z.enum([
  "gravity",
  "wind",
  "attraction",
  "explosion",
  "buoyancy",
  "vortex",
  "drag",
]);

export type ForceType = z.infer<typeof ForceTypeSchema>;

export const ShapeTypeSchema = z.enum([
  "circle",
  "box",
  "polygon",
  "capsule",
  "convex",
  "compound",
]);

export type ShapeType = z.infer<typeof ShapeTypeSchema>;

export const CollisionResponseSchema = z.enum([
  "collide",
  "sensor",
  "none",
]);

export type CollisionResponse = z.infer<typeof CollisionResponseSchema>;

// ═══════════════════════════════════════════════════════════════════════════
// Core Physics Types
// ═══════════════════════════════════════════════════════════════════════════

export const PhysicsVec2Schema = z.object({
  x: finiteNumber,
  y: finiteNumber,
}).strict();

export type PhysicsVec2 = z.infer<typeof PhysicsVec2Schema>;

export const PhysicsMaterialSchema = z.object({
  restitution: normalized01,
  friction: nonNegativeFinite.max(10), // Max reasonable friction
  surfaceVelocity: PhysicsVec2Schema.optional(),
}).strict();

export type PhysicsMaterial = z.infer<typeof PhysicsMaterialSchema>;

export const CollisionShapeSchema: z.ZodType<any> = z.lazy(() =>
  z.object({
    type: ShapeTypeSchema,
    radius: positiveFinite.max(10000).optional(),
    width: positiveFinite.max(10000).optional(),
    height: positiveFinite.max(10000).optional(),
    vertices: boundedArray(PhysicsVec2Schema, 1000).optional(), // Max 1000 vertices per polygon
    length: positiveFinite.max(10000).optional(),
    shapes: boundedArray(CollisionShapeSchema, 100).optional(), // Max 100 nested shapes
    offset: PhysicsVec2Schema.optional(),
    rotation: finiteNumber.optional(),
  }).strict().refine(
    (data) => {
      // Circle requires radius
      if (data.type === "circle" && data.radius === undefined) {
        return false;
      }
      // Box requires width and height
      if (data.type === "box" && (data.width === undefined || data.height === undefined)) {
        return false;
      }
      // Polygon requires vertices
      if (data.type === "polygon" && (!data.vertices || data.vertices.length < 3)) {
        return false;
      }
      // Capsule requires radius and length
      if (data.type === "capsule" && (data.radius === undefined || data.length === undefined)) {
        return false;
      }
      // Compound requires shapes
      if (data.type === "compound" && (!data.shapes || data.shapes.length === 0)) {
        return false;
      }
      return true;
    },
    { message: "Shape type requires appropriate properties" }
  )
);

export type CollisionShape = z.infer<typeof CollisionShapeSchema>;

export const CollisionFilterSchema = z.object({
  category: z.number().int().nonnegative().max(32), // Max 32 categories (bitmask)
  mask: z.number().int().nonnegative().max(4294967295), // Max 32-bit unsigned int
  group: z.number().int().max(32767).min(-32768), // 16-bit signed int range
}).strict();

export type CollisionFilter = z.infer<typeof CollisionFilterSchema>;

// ═══════════════════════════════════════════════════════════════════════════
// Rigid Body
// ═══════════════════════════════════════════════════════════════════════════

export const RigidBodyConfigSchema = z.object({
  id: entityId,
  layerId: entityId,
  type: BodyTypeSchema,
  mass: positiveFinite.max(1000000), // Max reasonable mass (1M kg)
  moment: positiveFinite.max(1000000000).optional(), // Max reasonable moment of inertia
  position: PhysicsVec2Schema,
  velocity: PhysicsVec2Schema.refine(
    (v) => Math.abs(v.x) < 10000 && Math.abs(v.y) < 10000,
    { message: "Velocity components must be within reasonable range" }
  ),
  angle: finiteNumber,
  angularVelocity: finiteNumber.max(1000).min(-1000), // Max reasonable angular velocity (rad/s)
  shape: CollisionShapeSchema,
  material: PhysicsMaterialSchema,
  filter: CollisionFilterSchema,
  response: CollisionResponseSchema,
  linearDamping: normalized01,
  angularDamping: normalized01,
  fixedRotation: z.boolean().optional(),
  bullet: z.boolean().optional(),
  canSleep: z.boolean(),
  sleepThreshold: positiveFinite.max(100), // Max reasonable sleep threshold
}).strict();

export type RigidBodyConfig = z.infer<typeof RigidBodyConfigSchema>;

export const ContactInfoSchema = z.object({
  bodyA: entityId,
  bodyB: entityId,
  point: PhysicsVec2Schema,
  normal: PhysicsVec2Schema.refine(
    (n) => {
      const length = Math.sqrt(n.x * n.x + n.y * n.y);
      return Math.abs(length - 1.0) < 0.01; // Normalized vector
    },
    { message: "Normal vector must be normalized" }
  ),
  depth: positiveFinite.max(10000),
}).strict().refine(
  (data) => data.bodyA !== data.bodyB,
  { message: "bodyA and bodyB must be different", path: ["bodyB"] }
);

export type ContactInfo = z.infer<typeof ContactInfoSchema>;

export const RigidBodyStateSchema = z.object({
  id: entityId,
  position: PhysicsVec2Schema,
  velocity: PhysicsVec2Schema,
  angle: finiteNumber,
  angularVelocity: finiteNumber,
  isSleeping: z.boolean(),
  contacts: boundedArray(ContactInfoSchema, 1000), // Max 1000 contacts per body
}).strict();

export type RigidBodyState = z.infer<typeof RigidBodyStateSchema>;

File: standalone/lattice-core/haskell/Lattice/Schemas/Physics/PhysicsSchema.hs
{-# LANGUAGE StrictData #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE OverloadedStrings #-}

{-|
Module      : Lattice.Schemas.Physics.PhysicsSchema
Description : Physics simulation types
Copyright   : (c) Lattice, 2026
License     : MIT

Physics simulation type enums and core types.

Source: ui/src/schemas/physics/physics-schema.ts
-}

module Lattice.Schemas.Physics.PhysicsSchema
  ( -- * Body Type
    BodyType(..)
  , bodyTypeFromText
  , bodyTypeToText
    -- * Joint Type
  , JointType(..)
  , jointTypeFromText
  , jointTypeToText
    -- * Force Type
  , ForceType(..)
  , forceTypeFromText
  , forceTypeToText
    -- * Shape Type
  , ShapeType(..)
  , shapeTypeFromText
  , shapeTypeToText
    -- * Collision Response
  , CollisionResponse(..)
  , collisionResponseFromText
  , collisionResponseToText
    -- * Physics Vec2
  , PhysicsVec2(..)
    -- * Physics Material
  , PhysicsMaterial(..)
    -- * Collision Filter
  , CollisionFilter(..)
  ) where

import GHC.Generics (Generic)
import Data.Text (Text)

-- ────────────────────────────────────────────────────────────────────────────
-- Body Type
-- ────────────────────────────────────────────────────────────────────────────

-- | Physics body type options
data BodyType
  = BodyStatic
  | BodyDynamic
  | BodyKinematic
  | BodyDormant
  | BodyAEmatic
  | BodyDead
  deriving stock (Eq, Show, Generic, Enum, Bounded)

bodyTypeFromText :: Text -> Maybe BodyType
bodyTypeFromText "static" = Just BodyStatic
bodyTypeFromText "dynamic" = Just BodyDynamic
bodyTypeFromText "kinematic" = Just BodyKinematic
bodyTypeFromText "dormant" = Just BodyDormant
bodyTypeFromText "AEmatic" = Just BodyAEmatic
bodyTypeFromText "dead" = Just BodyDead
bodyTypeFromText _ = Nothing

bodyTypeToText :: BodyType -> Text
bodyTypeToText BodyStatic = "static"
bodyTypeToText BodyDynamic = "dynamic"
bodyTypeToText BodyKinematic = "kinematic"
bodyTypeToText BodyDormant = "dormant"
bodyTypeToText BodyAEmatic = "AEmatic"
bodyTypeToText BodyDead = "dead"

-- ────────────────────────────────────────────────────────────────────────────
-- Joint Type
-- ────────────────────────────────────────────────────────────────────────────

-- | Physics joint type options
data JointType
  = JointPivot
  | JointSpring
  | JointDistance
  | JointPiston
  | JointWheel
  | JointWeld
  | JointBlob
  | JointRope
  deriving stock (Eq, Show, Generic, Enum, Bounded)

jointTypeFromText :: Text -> Maybe JointType
jointTypeFromText "pivot" = Just JointPivot
jointTypeFromText "spring" = Just JointSpring
jointTypeFromText "distance" = Just JointDistance
jointTypeFromText "piston" = Just JointPiston
jointTypeFromText "wheel" = Just JointWheel
jointTypeFromText "weld" = Just JointWeld
jointTypeFromText "blob" = Just JointBlob
jointTypeFromText "rope" = Just JointRope
jointTypeFromText _ = Nothing

jointTypeToText :: JointType -> Text
jointTypeToText JointPivot = "pivot"
jointTypeToText JointSpring = "spring"
jointTypeToText JointDistance = "distance"
jointTypeToText JointPiston = "piston"
jointTypeToText JointWheel = "wheel"
jointTypeToText JointWeld = "weld"
jointTypeToText JointBlob = "blob"
jointTypeToText JointRope = "rope"

-- ────────────────────────────────────────────────────────────────────────────
-- Force Type
-- ────────────────────────────────────────────────────────────────────────────

-- | Physics force type options
data ForceType
  = ForceGravity
  | ForceWind
  | ForceAttraction
  | ForceExplosion
  | ForceBuoyancy
  | ForceVortex
  | ForceDrag
  deriving stock (Eq, Show, Generic, Enum, Bounded)

forceTypeFromText :: Text -> Maybe ForceType
forceTypeFromText "gravity" = Just ForceGravity
forceTypeFromText "wind" = Just ForceWind
forceTypeFromText "attraction" = Just ForceAttraction
forceTypeFromText "explosion" = Just ForceExplosion
forceTypeFromText "buoyancy" = Just ForceBuoyancy
forceTypeFromText "vortex" = Just ForceVortex
forceTypeFromText "drag" = Just ForceDrag
forceTypeFromText _ = Nothing

forceTypeToText :: ForceType -> Text
forceTypeToText ForceGravity = "gravity"
forceTypeToText ForceWind = "wind"
forceTypeToText ForceAttraction = "attraction"
forceTypeToText ForceExplosion = "explosion"
forceTypeToText ForceBuoyancy = "buoyancy"
forceTypeToText ForceVortex = "vortex"
forceTypeToText ForceDrag = "drag"

-- ────────────────────────────────────────────────────────────────────────────
-- Shape Type
-- ────────────────────────────────────────────────────────────────────────────

-- | Collision shape type options
data ShapeType
  = ShapeCircle
  | ShapeBox
  | ShapePolygon
  | ShapeCapsule
  | ShapeConvex
  | ShapeCompound
  deriving stock (Eq, Show, Generic, Enum, Bounded)

shapeTypeFromText :: Text -> Maybe ShapeType
shapeTypeFromText "circle" = Just ShapeCircle
shapeTypeFromText "box" = Just ShapeBox
shapeTypeFromText "polygon" = Just ShapePolygon
shapeTypeFromText "capsule" = Just ShapeCapsule
shapeTypeFromText "convex" = Just ShapeConvex
shapeTypeFromText "compound" = Just ShapeCompound
shapeTypeFromText _ = Nothing

shapeTypeToText :: ShapeType -> Text
shapeTypeToText ShapeCircle = "circle"
shapeTypeToText ShapeBox = "box"
shapeTypeToText ShapePolygon = "polygon"
shapeTypeToText ShapeCapsule = "capsule"
shapeTypeToText ShapeConvex = "convex"
shapeTypeToText ShapeCompound = "compound"

-- ────────────────────────────────────────────────────────────────────────────
-- Collision Response
-- ────────────────────────────────────────────────────────────────────────────

-- | Collision response options
data CollisionResponse
  = ResponseCollide
  | ResponseSensor
  | ResponseNone
  deriving stock (Eq, Show, Generic, Enum, Bounded)

collisionResponseFromText :: Text -> Maybe CollisionResponse
collisionResponseFromText "collide" = Just ResponseCollide
collisionResponseFromText "sensor" = Just ResponseSensor
collisionResponseFromText "none" = Just ResponseNone
collisionResponseFromText _ = Nothing

collisionResponseToText :: CollisionResponse -> Text
collisionResponseToText ResponseCollide = "collide"
collisionResponseToText ResponseSensor = "sensor"
collisionResponseToText ResponseNone = "none"

-- ────────────────────────────────────────────────────────────────────────────
-- Physics Vec2
-- ────────────────────────────────────────────────────────────────────────────

-- | 2D vector for physics
data PhysicsVec2 = PhysicsVec2
  { pv2X :: !Double
  , pv2Y :: !Double
  }
  deriving stock (Eq, Show, Generic)

-- ────────────────────────────────────────────────────────────────────────────
-- Physics Material
-- ────────────────────────────────────────────────────────────────────────────

-- | Physics material properties
data PhysicsMaterial = PhysicsMaterial
  { pmRestitution :: !Double      -- 0-1
  , pmFriction :: !Double         -- >= 0
  , pmSurfaceVelocity :: !(Maybe PhysicsVec2)
  }
  deriving stock (Eq, Show, Generic)

-- ────────────────────────────────────────────────────────────────────────────
-- Collision Filter
-- ────────────────────────────────────────────────────────────────────────────

-- | Collision filtering
data CollisionFilter = CollisionFilter
  { cfCategory :: !Int       -- Max 32 categories
  , cfMask :: !Int           -- Bitmask
  , cfGroup :: !Int          -- 16-bit signed range
  }
  deriving stock (Eq, Show, Generic)
